### PR TITLE
[FIX] website: hide .s_popup itself too, not only its .modal child

### PR DESCRIPTION
**Pull request by rdeodoo**
- Web Editor facts: Since the refactoring of the editor done at [1], the cookies bar popup is receiving a `contenteditable=true` attribute, making it receive the chrome default `height` style for such elements.
The problem is that if this cookies bar's modal is hidden, the bar will be shown as a thin empty white bar.

- Website Builder / Popup Snippet / Cookies bar facts: A popup element is composed of a `.s_popup` parent containing the actual `.modal` BS modal. Our internal logic and events are hidding and showing this inner `.modal` modal element without considering its `.s_popup` parent. It means that when the `.modal` is hidden, its `.s_popup` parent is not touched and kept visible.
It might looks like it's not an issue as it would just be an empty element (its only child is hidden) but it leads to some issues as explained just above: an ugly white bar is shown.
Note that the cookies bar is nothing more than a `.s_popup` snippet.

- Web Editor facts 2: During the mentioned refactoring [1], they actually added some code to hide this bug once you were playing with the edit mode (mainly for when you clicked on the invisible panel element or before saving). But this code was actually not fixing the case when you just entered edit mode.

This commit simply remove that "edit only" web editor logic and add a new one in charge of simply synchronizing the `.s_popup` snippet visibility with its `.modal` BS modal in a public widget (which is obviously also executed in edit mode).

Finally:
- note that if there is no way through regular flow to arrive to the same result with a normal popup snippet, it is still concerned by this issue and you can reproduce it by moving it somewhere else in the DOM and/or simply adding it the `contenteditable=true` attribute.
- we already fixed that issue a few months/years ago, I couldn't find a commit related to that so I don't know if it was due to the same root cause
- in case one is wondering why we simply couldn't do that in the already existing `_onHideModal` of the `PopupWidget`, it is because that `hide.bs.modal` is not called when entering edit mode, because the `destroy` is destroying it before hidding the modal. The event is thus not fired. And we can't move the hide part before the destroy's super because otherwise, it would go through the normal hidden process which is creating a "seen" cookie for the popup as if it was closed manually.

[1]: https://github.com/odoo/odoo/commit/740168ce8d27da3d6a7156d2d79655a898394923

task-2754108

### CODE REVIEW

File: addons/website/static/src/snippets/s_popup/000.js [hunk 1]
```diff
@@ -6,6 +6,37 @@ const dom = require('web.dom');
...
+        // element (its only child is hidden) but it leads to some issues as for
+        // instance on chrome this div will have a forced `height` due to its
+        // `contenteditable=true` attribute. It will result in a ugly white bar.
+        // tl;dr: this is keeping those 2 elements visibility synchronized.
+        this.$target.on('shown.bs.modal.SharedPopupWidget', () => {
+            this.$target.show();
+        });
+        this.$target.on('hide.bs.modal.SharedPopupWidget', () => {
+            this.$target.hide();
+        });
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_popup/000.js:33]: Probably buggy: you use the "shown" and "hide" events while it should probably be "show" and "hidden". My guess is that with "shown" and "hide", you don't see the opening/closing effects no more (since you only make the external div visible when the opening animation is finished and hide it before the closing animation starts).

File: addons/website_mass_mailing/static/tests/tours/newsletter_popup.js [hunk 2]
```diff
@@ -21,7 +21,7 @@ tour.register('newsletter_popup_edition', {
     ...wTourUtils.clickOnSave(),
     {
         content: "Check the modal has been saved, closed",
-        trigger: '.o_newsletter_popup',
+        trigger: 'body:has(.o_newsletter_popup)',
         extra_trigger: 'body:not(.editor_enable)',
         run: function (actions) {
             const $modal = this.$anchor.find('.modal');
```
Comment (qsm-odoo) [inline addons/website_mass_mailing/static/tests/tours/newsletter_popup.js:27]: Seems scary to change this without adapting the `run` method (this.$anchor is now different and would find all modal even invisible ones).

File: addons/website_mass_mailing/static/tests/tours/newsletter_popup.js [hunk 3]
```diff
@@ -44,13 +44,12 @@ tour.register('newsletter_popup_use', {
...
     {
         content: "Check the modal is not yet opened and force it opened",
-        trigger: '.o_newsletter_popup',
+        trigger: 'body:has(.o_newsletter_popup)',
         run: function (actions) {
             const $modal = this.$anchor.find('.modal');
             if ($modal.is(':visible')) {
                 console.error('Modal is already opened...');
             }
-            $(document).trigger('mouseleave');
```
Comment (qsm-odoo) [inline addons/website_mass_mailing/static/tests/tours/newsletter_popup.js:53]: How does it work without this? :thinking: Maybe the modal is in delay mode and this is is the remaining of a test which tested the "on exit" mode?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
> in case one is wondering why we simply couldn't do that in the already
  existing `_onHideModal` of the `PopupWidget`, it is because that
  `hide.bs.modal` is not called when entering edit mode, because the
  `destroy` is destroying it before hiding the modal. The event is thus
  not fired. And we can't move the hide part before the destroy's super
  because otherwise, it would go through the normal hidden process which
  is creating a "seen" cookie for the popup as if it was closed
  manually.

Well that seems fragile :smile: But I guess relying on the fact `destroy`'s super is call first is ok indeed. Although, should not the part about resetting iframes' src be part of the destroy then?

File: addons/website_mass_mailing/static/tests/tours/newsletter_popup.js (hunk 2 above)
Comment (rdeodoo) [reply addons/website_mass_mailing/static/tests/tours/newsletter_popup.js]: Correct, I adapted it.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 4]
```diff
@@ -2568,29 +2568,6 @@ options.registry.CookiesBar = options.registry.SnippetPopup.extend({
...
-        // We hide the parent because contenteditable="true" would force the bar to stay visible in hidden mode.
-        this.$target.parent('#website_cookies_bar').hide();
-        this._super(...arguments);
-    },
-    /**
-     * @override
-     */
-    cleanForSave: function () {
-        this.$target.parent('#website_cookies_bar').show();
-    },
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/static/src/js/editor/snippets.options.js:2593]: Yup :grimacing:

File: addons/website/static/src/snippets/s_popup/000.js (hunk 1 above)
Comment (rdeodoo) [reply addons/website/static/src/snippets/s_popup/000.js]: Damn, indeed :facepalm:

File: addons/website/static/src/snippets/s_popup/000.js [hunk 5]
```diff
@@ -6,6 +6,37 @@ const dom = require('web.dom');
...
+        // and showing this inner `.modal` modal element without considering its
+        // `.s_popup` parent. It means that when the `.modal` is hidden, its
+        // `.s_popup` parent is not touched and kept visible.
+        // It might looks like it's not an issue as it would just be an empty
+        // element (its only child is hidden) but it leads to some issues as for
+        // instance on chrome this div will have a forced `height` due to its
+        // `contenteditable=true` attribute. It will result in a ugly white bar.
+        // tl;dr: this is keeping those 2 elements visibility synchronized.
+        this.$target.on('shown.bs.modal.SharedPopupWidget', () => {
+            this.$target.show();
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/static/src/snippets/s_popup/000.js]: Removed jquery, I also refactored the event listener and moved it through a `events` property of the Widget, not sure why I did it that way in the first place (well, I know, I implemented it differently then simplified it)

**Review by rdeodoo — COMMENTED** [review]
Applied changes here: https://github.com/odoo/odoo/compare/68f544a2d62693fbcfd7ffe4bd84f2670af29cc6..3d6b76d1919afac115d7e5b9d07e8150fecfda80

>  Although, should not the part about resetting iframes' src be part of the destroy then?

Not really, this is needed when one is closing the popup for the video to stop playing, and when he does that there is no destroy involved

File: addons/website/static/src/snippets/s_popup/000.js [hunk 6]
```diff
@@ -6,6 +6,37 @@ const dom = require('web.dom');
...
+        // element (its only child is hidden) but it leads to some issues as for
+        // instance on chrome this div will have a forced `height` due to its
+        // `contenteditable=true` attribute. It will result in a ugly white bar.
+        // tl;dr: this is keeping those 2 elements visibility synchronized.
+        this.$target.on('show.bs.modal.SharedPopupWidget', () => {
+            this.$target[0].classList.remove('d-none');
+        });
+        this.$target.on('hidden.bs.modal.SharedPopupWidget', () => {
+            this.$target[0].classList.add('d-none');
+        });
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_popup/000.js:33]: You bind something in the start but don't unbind it in the destroy actually :grimacing: So after you enter edit mode, each event handler is probably called twice when a modal opens/closes... that does not break anything I guess but should still be changed :+1: 

Also this may be fragile actually: when you save a page with a modal opened, the modal is first closed before save... but this handle which adds the d-none class is called after the closing animation is done (as it should). This may be random, I would not be surprise that sometimes the snippet is sometimes saved with the d-none class and sometimes not. Maybe you should explicitly add the d-none class on save? Although I guess that saved with or without the class, it will still work, best be consistent I think.

File: addons/website_mass_mailing/static/tests/tours/newsletter_popup.js [hunk 7]
```diff
@@ -44,13 +44,12 @@ tour.register('newsletter_popup_use', {
 }, [
     {
         content: "Check the modal is not yet opened and force it opened",
-        trigger: '.o_newsletter_popup',
-        run: function (actions) {
-            const $modal = this.$anchor.find('.modal');
+        trigger: 'body:has(.o_newsletter_popup)',
+        run: () => {
+            const $modal = $('.o_newsletter_popup .modal');
```
Comment (qsm-odoo) [inline addons/website_mass_mailing/static/tests/tours/newsletter_popup.js:49]: Careful that in 16.0/master, searching in the whole DOM won't work and this will be silently broken since the jquery style `if` will be false instead of crashing as it should.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
>>    Although, should not the part about resetting iframes' src be part of the destroy then?
>
> Not really, this is needed when one is closing the popup for the video to stop playing, and when he does that there is no destroy involved

:thinking: Open a modal with a video in it, enter in edit mode -> destroy is called :shrug: Maybe there is no bug and nothing to do but I don't understand your explanation :grimacing:

File: addons/website/static/src/snippets/s_popup/000.js (hunk 6 above)
Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_popup/000.js:33]: > You bind something in the start but don't unbind it in the destroy actually grimacing

Note: why not use the "events" property to declare those handlers actually?

File: addons/website_mass_mailing/static/tests/tours/newsletter_popup.js (hunk 7 above)
Comment (rdeodoo) [reply addons/website_mass_mailing/static/tests/tours/newsletter_popup.js]: Yep, this was not strong enough anymore.
I added a check to ensure we found one (and only one) such modal, that will prevent the test to silently fail if wrong or no element is found

**Review by rdeodoo — COMMENTED** [review]
Thanks, applied [here](https://github.com/odoo/odoo/compare/3d6b76d1919afac115d7e5b9d07e8150fecfda80..255fee757321af903bb8254df39b1c3c239458e4)

File: addons/website/static/src/snippets/s_popup/000.js (hunk 6 above)
Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_popup/000.js:33]: @rdeodoo 

> We really should think about creating a util method to bind events (when we can't use the events attribute ofc) that would unbind the event in destroy automatically IMO, this seems to be a frequent issue everyone is making, no?

No, it's very rare: what you want is to use the `events`  key which is exactly there for that.

> I did that, see [here](https://github.com/odoo/odoo/compare/72129cf7d5dfc19799a8ceb989c431970a45134f..ab27efd3f58dccba29581a6f27c5971e195e81aa), but it was not working, the event was never triggered

This is what should be investigated IMO, I don't see the reason why it would not work so it may be a bug worth checking... The system normally do exactly what you are doing: binding events using jquery's `on` etc :grimacing: 

> Sure, but not sure where would it be better? In the cleanForSave in the s_popup/options.js? Or the other possibility would be in the cleanForSave of the SnippetEditor [here](https://github.com/odoo/odoo/blame/331b9435c59936c604c50f46f448ca19929b79e3/addons/web_editor/static/src/js/editor/snippets.editor.js#L326) which is in charge of closing the popup when saving? With a check if this is a .s_popup?

No, in this code right here: in the `destroy` of this public widget IMO :+1: Kinda a corner case but technically, the snippet XML definition should have the d-none class... but that may hurt the drag and drop system so probably cannot be done even in master. But in the end, it's natural that this is part of the widget system: this is the one in charge of synchronizing the snippet div d-none class with the inner modal visibility.

Note: actually you also used `$target` in this public widget which is the deprecated/compatibility stuff. You should use `this.$el`/`this.el` only (but I'd really like to know why the `events` prop does not work first anyway)

**Comment by qsm-odoo** [comment]
> I meant that we can't move the iframe src reset inside the destroy method

Of course, I was not suggesting to move it but to call it there too, if it makes sense

File: addons/website/static/src/snippets/s_popup/000.js (hunk 6 above)
Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_popup/000.js]: >> We really should think about creating a util method to bind events (when we can't use the events attribute ofc) that would unbind the event in destroy automatically IMO, this seems to be a frequent issue everyone is making, no?
>
>No, it's very rare: what you want is to use the events key which is exactly there for that.
>
>> I did that, see [here](https://github.com/odoo/odoo/compare/72129cf7d5dfc19799a8ceb989c431970a45134f..ab27efd3f58dccba29581a6f27c5971e195e81aa), but it was not working, the event was never triggered
>
> This is what should be investigated IMO, I don't see the reason why it would not work so it may be a bug worth checking... The system normally do exactly what you are doing: binding events using jquery's on etc grimacing

I investigated, everything works fine on my end with `events`, so I push-forced with that version
